File: bank_client.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"net"
	"os"
	"strconv"
	"sync"
	"time"
)

const (
	HOST = "localhost"
	PORT = 9000
)

func sendAndRecv(msg string) string {
	conn, err := net.Dial("tcp", fmt.Sprintf("%s:%d", HOST, PORT))
	if err != nil {
		return fmt.Sprintf("ERROR|%v", err)
	}
	defer conn.Close()

	// Identificarse como cliente bancario
	fmt.Fprintf(conn, "CLIENT_BANK|bank_go\n")

	// Leer mensaje de bienvenida
	scanner := bufio.NewScanner(conn)
	scanner.Scan()

	// Enviar comando
	fmt.Fprintf(conn, "%s\n", msg)

	// Leer respuesta
	if scanner.Scan() {
		return scanner.Text()
	}

	return "ERROR|NoResponse"
}

func createAccounts(n int, initial float64) {
	msg := fmt.Sprintf("CREATE_ACCOUNTS|%d|%.2f", n, initial)
	resp := sendAndRecv(msg)
	fmt.Println(resp)
}

func consultAccount(aid int) {
	msg := fmt.Sprintf("CONSULTAR_CUENTA|%d", aid)
	resp := sendAndRecv(msg)
	fmt.Println(resp)
}

func transfer(from, to int, amount float64) {
	msg := fmt.Sprintf("TRANSFERIR_CUENTA|%d|%d|%.2f", from, to, amount)
	resp := sendAndRecv(msg)
	fmt.Println(resp)
}

func createLoan(accountID int, amount float64) {
	msg := fmt.Sprintf("CREAR_PRESTAMO|%d|%.2f|%.2f", accountID, amount, amount)
	resp := sendAndRecv(msg)
	fmt.Println(resp)
}

func payLoan(accountID, loanID int, amount float64) {
	msg := fmt.Sprintf("PAGAR_PRESTAMO|%d|%d|%.2f", accountID, loanID, amount)
	resp := sendAndRecv(msg)
	fmt.Println(resp)
}

func loanStatus(accountID int) {
	msg := fmt.Sprintf("ESTADO_PAGO_PRESTAMO|%d", accountID)
	resp := sendAndRecv(msg)
	fmt.Println(resp)
}

func arqueo() {
	msg := "ARQUEO"
	resp := sendAndRecv(msg)
	fmt.Println(resp)
}

func stressTest(numTx, maxAcc int, maxAmt float64) {
	fmt.Printf("Iniciando stress test: %d transacciones con cuentas 1-%d\n", numTx, maxAcc)

	var wg sync.WaitGroup
	rand.Seed(time.Now().UnixNano())

	for i := 0; i < numTx; i++ {
		wg.Add(1)
		go func(txNum int) {
			defer wg.Done()

			from := rand.Intn(maxAcc) + 1
			to := rand.Intn(maxAcc) + 1
			for to == from {
				to = rand.Intn(maxAcc) + 1
			}

			amount := rand.Float64() * maxAmt
			transfer(from, to, amount)

			// Delay aleatorio
			time.Sleep(time.Duration(rand.Intn(100)) * time.Millisecond)
		}(i)
	}

	wg.Wait()
	fmt.Println("Stress test completado")
}

func main() {
	if len(os.Args) < 2 {
		printUsage()
		return
	}

	cmd := os.Args[1]

	switch cmd {
	case "create":
		createAccounts(10000, 100.0)

	case "consult":
		if len(os.Args) < 3 {
			fmt.Println("Uso: bank_client consult <id>")
			return
		}
		id, _ := strconv.Atoi(os.Args[2])
		consultAccount(id)

	case "transfer":
		if len(os.Args) < 5 {
			fmt.Println("Uso: bank_client transfer <from> <to> <amount>")
			return
		}
		from, _ := strconv.Atoi(os.Args[2])
		to, _ := strconv.Atoi(os.Args[3])
		amount, _ := strconv.ParseFloat(os.Args[4], 64)
		transfer(from, to, amount)

	case "stress":
		numTx := 200
		maxAcc := 10000
		if len(os.Args) > 2 {
			numTx, _ = strconv.Atoi(os.Args[2])
		}
		if len(os.Args) > 3 {
			maxAcc, _ = strconv.Atoi(os.Args[3])
		}
		stressTest(numTx, maxAcc, 50.0)

	case "create_loan":
		if len(os.Args) < 4 {
			fmt.Println("Uso: bank_client create_loan <account_id> <amount>")
			return
		}
		accountID, _ := strconv.Atoi(os.Args[2])
		amount, _ := strconv.ParseFloat(os.Args[3], 64)
		createLoan(accountID, amount)

	case "pay_loan":
		if len(os.Args) < 5 {
			fmt.Println("Uso: bank_client pay_loan <account_id> <loan_id> <amount>")
			return
		}
		accountID, _ := strconv.Atoi(os.Args[2])
		loanID, _ := strconv.Atoi(os.Args[3])
		amount, _ := strconv.ParseFloat(os.Args[4], 64)
		payLoan(accountID, loanID, amount)

	case "loan_status":
		if len(os.Args) < 3 {
			fmt.Println("Uso: bank_client loan_status <account_id>")
			return
		}
		accountID, _ := strconv.Atoi(os.Args[2])
		loanStatus(accountID)

	case "arqueo":
		arqueo()

	default:
		printUsage()
	}
}

func printUsage() {
	fmt.Println("Uso:")
	fmt.Println("  bank_client create")
	fmt.Println("  bank_client consult <id>")
	fmt.Println("  bank_client transfer <from> <to> <amount>")
	fmt.Println("  bank_client stress [num_tx] [max_acc]")
	fmt.Println("  bank_client create_loan <account_id> <amount>")
	fmt.Println("  bank_client pay_loan <account_id> <loan_id> <amount>")
	fmt.Println("  bank_client loan_status <account_id>")
	fmt.Println("  bank_client arqueo")
}

File: bank_client.py
#!/usr/bin/env python3

import socket, sys, time, random, threading

HOST = "localhost"
PORT = 9000


def send_and_recv(msg):
    """Envía un mensaje al servidor central y devuelve la respuesta."""
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.connect((HOST, PORT))
        s.sendall(b"CLIENT_BANK|bank1\n")
        s.recv(4096)  # Mensaje de bienvenida
        s.sendall((msg + "\n").encode())
        data = s.recv(8192).decode()
        return data.strip()
    except Exception as e:
        return f"ERROR|{e}"
    finally:
        try:
            s.close()
        except:
            pass


def create_accounts(n, initial):
    """Crea 'n' cuentas con saldo inicial 'initial'."""
    resp = send_and_recv(f"CREATE_ACCOUNTS|{n}|{initial}")
    print(resp)


def consult_account(aid):
    """Consulta el estado de la cuenta con id 'aid'."""
    resp = send_and_recv(f"CONSULTAR_CUENTA|{aid}")
    print(resp)


def transfer(a, b, m):
    """Transfiere un monto exacto de una cuenta a otra."""
    msg = f"TRANSFERIR_CUENTA|{a}|{b}|{m}"
    resp = send_and_recv(msg)
    print(resp)


def create_loan(account_id, amount):
    """Crea un préstamo para una cuenta."""
    msg = f"CREAR_PRESTAMO|{account_id}|{amount}|{amount}"
    resp = send_and_recv(msg)
    print(resp)


def pay_loan(account_id, loan_id, amount):
    """Paga una parte del préstamo."""
    msg = f"PAGAR_PRESTAMO|{account_id}|{loan_id}|{amount}"
    resp = send_and_recv(msg)
    print(resp)


def loan_status(account_id):
    """Consulta el estado de préstamos de una cuenta."""
    msg = f"ESTADO_PAGO_PRESTAMO|{account_id}"
    resp = send_and_recv(msg)
    print(resp)


def arqueo():
    """Realiza un arqueo total del sistema."""
    msg = "ARQUEO"
    resp = send_and_recv(msg)
    print(resp)


def stress_test(num_tx=100, max_acc=100, max_amt=100):
    """Realiza múltiples transferencias aleatorias concurrentes (modo stress test)."""

    def worker_tx(i):
        a = random.randint(1, max_acc)
        b = random.randint(1, max_acc)
        while b == a:
            b = random.randint(1, max_acc)
        amt = round(random.random() * max_amt, 2)
        transfer(a, b, amt)
        time.sleep(random.random() * 0.1)

    threads = []
    for i in range(num_tx):
        t = threading.Thread(target=worker_tx, args=(i,))
        t.start()
        threads.append(t)
    for t in threads:
        t.join()


if __name__ == "__main__":
    if len(sys.argv) >= 2 and sys.argv[1] == "create":
        # Por defecto crea 10000 cuentas con 100.0 cada una
        create_accounts(10000, 100.0)

    elif len(sys.argv) >= 2 and sys.argv[1] == "consult":
        if len(sys.argv) < 3:
            print("Uso: bank_client.py consult <id>")
        else:
            consult_account(sys.argv[2])

    elif len(sys.argv) >= 2 and sys.argv[1] == "transfer":
        if len(sys.argv) < 5:
            print("Uso: bank_client.py transfer <from> <to> <amt>")
        else:
            try:
                a = int(sys.argv[2])
                b = int(sys.argv[3])
                m = float(sys.argv[4])
                transfer(a, b, m)
            except ValueError:
                print("Argumentos inválidos. Uso: bank_client.py transfer <from> <to> <amt>")

    elif len(sys.argv) >= 2 and sys.argv[1] == "stress":
        num = int(sys.argv[2]) if len(sys.argv) > 2 else 200
        maxacc = int(sys.argv[3]) if len(sys.argv) > 3 else 10000
        stress_test(num, maxacc, 50)

    elif len(sys.argv) >= 2 and sys.argv[1] == "create_loan":
        if len(sys.argv) < 4:
            print("Uso: bank_client.py create_loan <account_id> <amount>")
        else:
            create_loan(sys.argv[2], float(sys.argv[3]))

    elif len(sys.argv) >= 2 and sys.argv[1] == "pay_loan":
        if len(sys.argv) < 5:
            print("Uso: bank_client.py pay_loan <account_id> <loan_id> <amount>")
        else:
            pay_loan(sys.argv[2], int(sys.argv[3]), float(sys.argv[4]))

    elif len(sys.argv) >= 2 and sys.argv[1] == "loan_status":
        if len(sys.argv) < 3:
            print("Uso: bank_client.py loan_status <account_id>")
        else:
            loan_status(sys.argv[2])

    elif len(sys.argv) >= 2 and sys.argv[1] == "arqueo":
        arqueo()

    else:
        print("Uso:")
        print("  bank_client.py create")
        print("  bank_client.py consult <id>")
        print("  bank_client.py transfer <from> <to> <amt>")
        print("  bank_client.py stress [num_tx] [max_acc]")
        print("  bank_client.py create_loan <account_id> <amount>")
        print("  bank_client.py pay_loan <account_id> <loan_id> <amount>")
        print("  bank_client.py loan_status <account_id>")
        print("  bank_client.py arqueo")
